soccer/gameplay/evaluation/touchpass_positioning.py

```diff
@@ -0,0 +1,100 @@
+
+## The Touchpass Positioner finds the best location within a rectangle
+#
+# Example usage:
+# tpos = TouchpassPositioner()
+# best_pt = tpos.evaluate(ball.pos, robot2.pos)
+class TouchpassPositioner:
+
+    def __init__(self):
+        self.debug = False
+
+    # Defaults to False
+    # if True, uses the system state drawing methods to draw Windows
+    @property
+    def debug(self):
+        return self._debug
+    @debug.setter
+    def debug(self, value):
+        self._debug = value
+
+    ## Returns a robocup.Rect object that is the default location to be evaluated
+    #
+    # Takes a current ball position/initial kick position (robocup.Point)
+    def generate_default_rectangle(self, kick_point):
+        offset_from_edge = 0.3
+
+        if kick_point.x > 0:
+            # Ball is on right side of field
+            toReturn = robocup.Rect(robocup.Point(0, min(constants.Field.Length - offset_from_edge, main.ball().pos.y - 0.3)),
+                    robocup.Point(-constants.Field.Width / 2 + offset_from_edge, min(constants.Field.Length * 3 / 4, main.ball().pos.y - 2)))
+        else:
+            # Ball is on left side of field
+            toReturn = robocup.Rect(robocup.Point(0, min(constants.Field.Length - offset_from_edge, main.ball().pos.y - 0.3)),
+                    robocup.Point(constants.Field.Width / 2 - offset_from_edge, min(constants.Field.Length * 3 / 4, main.ball().pos.y - 2)))
+        return toReturn
+
+    ## Returns a list of robocup.Point object that represent candidate points. Takes in a robocup.Rect
+    def get_points_from_rect(self, rect, threshold=0.50):
+        outlist = []
+        currentx = rect.min_x()
+        currenty = rect.max_y()
+
+        # Loop through from top left to bottom right
+
+        while currenty > rect.min_y():
+            while currentx < rect.max_x():
+                candiate = robocup.Point(currentx, currenty)
+                if not constants.Field.TheirGoalShape.contains_point(candiate):
+                    outlist.extend([candiate])
+                currentx = currentx + threshold
+            currenty = currenty - threshold
+            currentx = rect.min_x()
+        return outlist
+
+    ## Evaluates a single point, and returns the probability of it making it.
+    def eval_single_point(self, kick_point, receive_point):
+        if kick_point is None:
+            kick_point = main.ball().pos
+        currentChance = evaluation.passing.eval_pass(kick_point, receive_point)
+        # TODO dont only aim for center of goal. Waiting on window_evaluator returning a probability.
+        targetPoint = constants.Field.TheirGoalCenter
+        currentChance = currentChance * evaluation.passing.eval_pass(receive_point, targetPoint)
+        return currentChance
+
+
+    ## Finds the best receive point for a bounce-pass.
+    #
+    # Takes in an initial kick point and an optional evaluation zone.
+    def eval_best_receive_point(self, kick_point, evaluation_zone=None):
```

Comment:
The default rectangle is dependent on the ball's current position, and using that would cause the rectangle to only be generated once. I did this to make sure that the rectangle would be regenerated every time. Don't want to run into that problem we were having with coordinated_pass again :smile: